Added support for Dictionaries

Richard-Degenne opened the pull request:
This PR adds basic support for TextRazor's dictionary features.

It adds 4 new endpoints to the API :

- Creation of a dictionary;
- Deletion of a dictionary;
- Creation of dictionary entries;
- Deletion of a dictionary entry.

The PR also revamps the way requests are handled, since until now, only the analysis API was considered.

andhapp commented:
Hello @Richard-Degenne,

Nice to meet you!

Thanks for submitting a PR. I will look through it in a couple of days.

Richard-Degenne commented:
Hi @andhapp,

I added support for another endpoint of the dictionary API.

Richard-Degenne commented:
Any input on the current state of this PR?

andhapp commented:
Sorry man!

It's been failing on Travis. I will have a look at it tonight. It must be something in the setup. I have glanced through the PR and it seems okay. Will look through it again just to make sure. 

Thanks for your patience!

Richard-Degenne commented:
Yes, the Travis error seems unrelated, but I can't say for sure.

> Thanks for your patience!

No worries. :)

andhapp commented on README.md:
````diff
@@ -73,6 +73,48 @@ TextRazor.phrases('api_key', 'text')
+#### Adding entries to a dictionary
+
+```
+entry = TextRazor::DictionaryEntry(id: 'my-entry', text: 'Text to be matched')
````
`DictionaryEntry` and `Dictionary` should be internal classes, IMHO. The API interface should be something like this:

```
client = TextRazor::Client.new('api_key')
client.create_dictionary('my-dictionary') 
# This can have keyword arguments but I'd not do it if it's not anywhere else in the library, makes it inconsistent
client.create_dictionary_entries('my-dictionary', {id: 'my-entry', text: 'Text to be matched'})
```

These method calls `create_dictionary` and `create_dictionary_entries` should return the actual instances of `Dictionary` and `DictionaryEntry` but there is no need for the consumer to know how to create these instances. `client` hides that behind a facade and for a good reason.

Richard-Degenne replied:
Agreed. I'll rework these parts.

andhapp commented on lib/textrazor/api_client.rb:
```diff
@@ -0,0 +1,36 @@
+require 'rest_client'
```
What's the main reason to separate this out into an `api_client` class? `

Richard-Degenne replied:
Agreed. I'll rework this part.

andhapp commented on lib/textrazor/request.rb:
```diff
@@ -21,18 +16,59 @@ class Request
 
     def self.build_query(text, options)
-      query = {"text" => text, "apiKey" => options.delete(:api_key)}
+      query = {"text" => text}
```
Removing it from here means this is not duplicated across 4 or 5 different methods. Any particular reason to do this?

Richard-Degenne replied:
The API key must be passed as a `x-textrazor-key` header.

As of today, this is not an issue because the only endpoint supported by the gem is a `POST` request, but adding support for dictionaries means supporting `GET`, `PUT` and `DELETE` requests.

andhapp commented on lib/textrazor/api_response.rb:
```diff
@@ -0,0 +1,54 @@
+
+module TextRazor
+
+  class ApiResponse
```
Similar to `ApiRequest`, `Response` class was meant to take care of this responsibility. Any particular reason to do this.

Richard-Degenne replied:
The `Response` class is heavily oriented towards parsing responses to analysis requests.

The more general `ApiResponse` can handle any kind of response from TextRazor. (Side note: `Response` inherits from `ApiResponse`)

andhapp commented:
Hey @Richard-Degenne,

I had a look at the PR and left some comments on it.

Richard-Degenne commented:
Thanks for the review! I'm a bit snowed under at the moment, but I'll look into your remarks as soon as I have a moment.

Richard-Degenne commented:
Hi @andhapp,

It's been a while, but I intend to work on this PR in the coming days. I'll answer your reviews and rework some things here and there.

Commit:
Added source_id to entities

Commit:
Added Dictionary model

Commit:
Revamped API calls

Commit:
Added dictionary creation

Commit:
Added Dictionary deletion

Commit:
Added DictionaryEntry

Commit:
Added dictionary entry creation

Commit:
Added dictionary entry deletion

Commit:
Added dictionary entry retrieval

Richard-Degenne commented:
Hi @andhapp,

I rebased the branch onto `master`, took your remarks into consideration and cleaned my commit history.

andhapp commented:
Thanks. I've reviewed your other PR. Will look at this one next.

Richard-Degenne commented:
Hi @andhapp,

Since the other PR got merged (thanks for that, btw), I rebased this one onto `master` and fixed the conflicts. Now, the dictionary support plays nice with the Europe endpoint support.

Richard-Degenne commented:
Hi @andhapp,

Any input on the current state of this PR? :)

Richard-Degenne commented:
Hi @andhapp,

Allow to bump this PR. I understand you have some problems with GitHub notifications, so maybe this one will get through. :stuck_out_tongue:

Commit:
Removed custom accept-encoding headers
